python/tank/descriptor/io_descriptor/base.py

```diff
@@ -211,93 +212,94 @@ def _find_latest_tag_by_pattern(self, version_numbers, pattern):
-            #       3: {}
-            #       }
-            #   }
-            # }
-            #
-            current = versions
-            for number in version_split:
-                if number not in current:
-                    current[number] = {}
-                current = current[number]
-
-        # now search for the latest version matching our pattern
-        if not re.match("^v([0-9]+|x)(.([0-9]+|x)){2,}$", pattern):
-            raise TankDescriptorError("Cannot parse version expression '%s'!" % pattern)
-
-        # split our pattern, beware each part is a string (even integers)
-        version_split = re.findall("([0-9]+|x)", pattern)
-        if 'x' in version_split:
-            # check that we don't have an incorrect pattern using x
-            # then a digit, eg. v4.x.2
-            if re.match("^v[0-9\.]+[x\.]+[0-9\.]+$", pattern):
-                raise TankDescriptorError(
-                    "Incorrect version pattern '%s'. "
-                    "There should be no digit after a 'x'." % pattern
-                )
-
-        current = versions
-        version_to_use = None
-        # process each digit in the pattern
-        for version_digit in version_split:
-            if version_digit == 'x':
-                # replace the 'x' by the latest at this level
-                version_digit = max(current.keys(), key=int)
-            version_digit = int(version_digit)
-            if version_digit not in current:
-                # no matches
+        # Validate the pattern: we expect at least v<digit or x>.<digit or x>.<digit or x>
+        # - we can then have more .<digit or x>
+        # - we can then have a -<topic name>, optionally followed by multiple .<alphanumeric or _>
+        if not re.match("^v(\d+|x)(\.(\d+|x)){2,}(-\w+(\.\w+)*)?$", pattern):
+            raise TankDescriptorError(
+                "Cannot parse version expression '%s', it must follow a vx.x.x[-topic] pattern "
+                "where 'x' can either be a digit or the letter 'x', and 'topic' is an "
+                "optional arbitrary string." % pattern
+            )
+        # Split the pattern either with "." or "-", keep the separator in the result
+        # with the grouping in the regexp
+        version_split = re.split("(\.|-)", pattern)
+        num_tokens = len(version_split)
+        # Special case for the first token for which the wildcard is "vx" and not
+        # "x"
+        has_v_prefix = False
+        version_regexp = version_split[0]
+        if version_split[0] == "vx":
+            has_v_prefix = True
+            version_split[0] = "x"
+        # Build a base version string that versions will have to match. The base
+        # version contains all tokens until we find a "x".
+        # Separators are kept by the split being done above, so actual tokens are
+        # every other entry and we need to only consider even entries, hence the '2'
+        # step
+        for i in range(0, num_tokens, 2):
```

Comment:
The last `x` will be treated exactly like the (major, minor, patch) wildcards, so would match `v1.2.3-foo`, `v1.2.3-foo.rc.1` and `v1.2.3-bar` or `v1.2.3-whatever.you.fancy`. But, it would not match `v1.2.3` as we check that we have at least the number of tokens which are specified in the pattern (4 tokens in vx.x.x-x and only 3 in v1.2.3). Does that seem right?